Spigot-Server-Patches/0619-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,310 @@
++import net.minecraft.server.BaseBlockPosition;
++import net.minecraft.server.MinecraftKey;
++import net.minecraft.server.PacketPlayOutCustomSoundEffect;
++import net.minecraft.server.Vec3D;
++import org.bukkit.Location;
++import org.bukkit.Sound;
++import org.bukkit.SoundCategory;
++import org.bukkit.World;
++import org.bukkit.craftbukkit.entity.CraftPlayer;
++import org.bukkit.entity.Player;
++import org.bukkit.util.NumberConversions;
++
++import java.util.ArrayList;
++import java.util.Collection;
++import java.util.List;
++
++public class PaperSoundBuilder implements SoundBuilder {
++
++    private Sound sound;
++    private SoundCategory category;
++    private Collection<Player> receivers;
++    private Location location;
++    private float volume = 1F;
++    private float pitch = 1F;
++
++    public PaperSoundBuilder(Sound sound, Location location) {
++        this.sound = sound;
++        this.location = location;
++    }
++
++    public PaperSoundBuilder(Sound sound, Location location, float volume) {
++        this.sound = sound;
++        this.location = location;
++        this.volume = volume;
++    }
++
++    public PaperSoundBuilder(Sound sound, Location location, Float volume, float pitch) {
++        this.sound = sound;
++        this.location = location;
++        this.volume = volume;
++        this.pitch = pitch;
++    }
++
++    public PaperSoundBuilder(Sound sound, Location location, SoundCategory category) {
++        this.sound = sound;
++        this.location = location;
++        this.category = category;
++    }
++
++    public PaperSoundBuilder(Sound sound, Location location, SoundCategory category, float volume) {
++        this.sound = sound;
++        this.location = location;
++        this.category = category;
++        this.volume = volume;
++    }
++
++    public PaperSoundBuilder(Sound sound, Location location, SoundCategory category, float volume, float pitch) {
++        this.sound = sound;
++        this.location = location;
++        this.category = category;
++        this.volume = volume;
++        this.pitch = pitch;
++    }
++
++    @Override
++    public PaperSoundBuilder play() {
++        if (this.category == null) {
++            throw new IllegalStateException("Please specify category for this sound");
++        }
++        if (receivers == null || receivers.isEmpty()) {
++            location.getWorld().playSound(location, sound, category, volume, pitch);
++        } else {
++            PacketPlayOutCustomSoundEffect customSoundEffect = new PacketPlayOutCustomSoundEffect(MinecraftKey.a(sound.getKey().getKey()), net.minecraft.server.SoundCategory.valueOf(category.name()), Vec3D.a(new BaseBlockPosition(location.getX(), location.getY(), location.getZ())), volume, pitch);
++            for (Player player : receivers) {
++                ((CraftPlayer) player).getHandle().playerConnection.sendPacket(customSoundEffect);
++            }
++        }
++        return this;
++    }
```

Comment:
```suggestion
+    @Override
+    public void play() {
+        if (receivers == null || receivers.isEmpty()) {
+            location.getWorld().playSound(location, sound, category, volume, pitch);
+        } else {
+            PacketPlayOutCustomSoundEffect customSoundEffect = new PacketPlayOutCustomSoundEffect(MinecraftKey.a(sound.getKey().getKey()), net.minecraft.server.SoundCategory.valueOf(category.name()), Vec3D.a(new BaseBlockPosition(location.getX(), location.getY(), location.getZ())), volume, pitch);
+            for (Player player : receivers) {
+                ((CraftPlayer) player).getHandle().playerConnection.sendPacket(customSoundEffect);
+            }
+        }
+    }
```

It's the end of the Builder. So it makes no sense doing something else afterwards. If someone wants to reuse it he should save the Builder into a variable before.